iceoryx_hoofs/include/iceoryx_hoofs/posix_wrapper/thread.hpp

```diff
@@ -28,10 +33,58 @@ constexpr uint64_t MAX_THREAD_NAME_LENGTH = 15U;
 
 using ThreadName_t = cxx::string<MAX_THREAD_NAME_LENGTH>;
 
-void setThreadName(pthread_t thread, const ThreadName_t& name) noexcept;
-ThreadName_t getThreadName(pthread_t thread) noexcept;
+enum class ThreadError
+{
+    EMPTY_CALLABLE,
+    INSUFFICIENT_MEMORY,
+    INSUFFICIENT_PERMISSIONS,
+    INSUFFICIENT_RESOURCES,
+    INVALID_ATTRIBUTES,
+    UNDEFINED
+};
+
+class Thread
+{
+  public:
+    using callable_t = cxx::function<void()>;
+
+    Thread(const Thread&) = delete;
+    Thread& operator=(const Thread&) = delete;
+    Thread(Thread&& other) = delete;
+    Thread& operator=(Thread&& rhs) = delete;
+
+    ~Thread() noexcept;
+
+    // @todo set name during creation and remove method
+    void setThreadName(const ThreadName_t& name) noexcept;
+
+    ThreadName_t getThreadName() noexcept;
+
+    friend class ThreadBuilder;
+    friend class cxx::optional<Thread>;
+
+  private:
+    Thread() noexcept = default;
+
+    static ThreadError errnoToEnum(const int errnoValue) noexcept;
+
+    static void* startRoutine(void* callable);
+
+    pthread_t m_threadHandle;
+    callable_t m_callable;
+    bool m_isJoinable{true};
+};
+
+class ThreadBuilder
+{
+  public:
+    cxx::expected<ThreadError> create(cxx::optional<Thread>& uninitializedThread,
+                                      const cxx::function<void()>& callable) noexcept;
```

Comment:
@FerdinandSpitzschnueffler and this is fine for now but in the mid to long term we might have to rethink this. Once it is possible to set the stack address, there are additional options how this could be handled.